### DISCUSSION TITLE
How to correctly build a yarn plugin?

### DISCUSSION THREAD

**Discussion by protango** [yarnpkg/berry#2872] [Q&A] [answered]
I'm trying to modify the `workspace-tools` plugin in my own fork of Berry (which is up to date with master), however if I go to the packages/plugin-workspace-tools folder and run `yarn run update-local`, it produces a completely different output file than the one currently committed.
In addition, if I try to install this outputted file in a different yarn repository, I get the following error when I attempt to use it:

```
$ yarn workspaces focus a
Usage Error: This plugin cannot access the package referenced via typanion which is neither a builtin, nor an exposed entry (when initializing @yarnpkg/plugin-workspace-tools, defined in /REDACTED/.yarnrc.yml)

Yarn Package Manager - 2.4.1

  $ yarn <command>

You can also print more details about any of these commands by calling them 
after adding the `-h,--help` flag right after the command name.
```

Am I doing something wrong with the `@yarnpkg/builder`? I've also tried using it to create a new plugin outside of the Berry repo, but had similar issues where it did not produce a valid output file.

**Comment by paul-soporan** [comment, marked as answer]
In our repository:
- `packages/*/bin` contains the committed bundles that we want to mark as latest - they are only updated when we make a release and they are currently on 2.x as 3.x is still in RC
- `packages/*/bundles` contains the locally generated bundles based on the current branch. They aren't committed and they are on 3.x since they use the current branch.
- You can't use 3.x plugins in 2.x projects.

> I've also tried using it to create a new plugin outside of the Berry repo, but had similar issues where it did not produce a valid output file.

Make sure you use the 3.x versions of the builder to create 3.x plugins and the 2.x versions of the builder to create 2.x plugins. 3.x plugins aren't compatible with 2.x, but 2.x plugins are mostly compatible with 3.x.

**Reply by protango** [reply]
Ah that makes more sense now, thanks! Do you think it might be worth adding a note on the builder readme to explain this?

**Reply by paul-soporan** [reply]
Sounds like a good idea :+1:
